### PR TITLE
prov/verbs: Simplify DGRAM code

### PR DESCRIPTION
**Pull request by dmitrygladkov**
This patch merges common functionalities of MSG and DGRAM EPs
into single code. Also re-implements DGRAM AV functionality
to make it more lightweight.

Now verbs/DGRAM requires support of FI_MSG_PREFIX from application.

The support of Completion counters and Event Queue were removed for verbs/DGRAM

Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>

### CODE REVIEW

**Comment by dmitrygladkov** [comment]
@shefty @a-ilango I'm sorry for a lot of code removed/rewritten/added within the single commit. But it's really hard to split them into different commits.
Can you take a look and provide feedback? This patch significantly improve readability of verbs provider code and significantly improve performance of verbs/DGRAM (measured on fi_pingpong test) -- that's all we love :)
@aingerson I think we may try to test RxD over verbs/DGRAM.

**Comment by dmitrygladkov** [comment]
image


Raw performance data collected with fi_pingpong:

verbs/DGRAM #4379 |   |   |  
-- | -- | -- | --
bytes | MB/sec | usec/xfer | Mxfers/sec
1 | 0.78 | 1.28 | 0.78
2 | 1.57 | 1.27 | 0.79
4 | 3.15 | 1.27 | 0.79
8 | 6.3 | 1.27 | 0.79
16 | 12.58 | 1.27 | 0.79
32 | 25.11 | 1.27 | 0.78
64 | 49.39 | 1.3 | 0.77
128 | 96.15 | 1.33 | 0.75
256 | 145.83 | 1.76 | 0.57
512 | 233.26 | 2.19 | 0.46
1k | 530.86 | 1.93 | 0.52
2k | 945.6 | 2.17 | 0.46
4k | 1554.65 | 2.63 | 0.38
verbs/DGRAM   master |   |  
bytes | MB/sec | usec/xfer | Mxfers/sec
1 | 0.65 | 1.53 | 0.65
2 | 1.31 | 1.53 | 0.65
4 | 2.63 | 1.52 | 0.66
8 | 5.24 | 1.53 | 0.65
16 | 10.49 | 1.53 | 0.66
32 | 21.06 | 1.52 | 0.66
64 | 41.22 | 1.55 | 0.64
128 | 81.2 | 1.58 | 0.63
256 | 128.53 | 1.99 | 0.5
512 | 210.47 | 2.43 | 0.41
1k | 504.3 | 2.03 | 0.49
2k | 896.58 | 2.28 | 0.44
4k | 1484.27 | 2.76 | 0.36
verbs/MSG |   |   |  
bytes | MB/sec | usec/xfer | Mxfers/sec
1 | 0.64 | 1.55 | 0.64
2 | 1.29 | 1.55 | 0.65
4 | 2.58 | 1.55 | 0.65
8 | 5.17 | 1.55 | 0.65
16 | 10.33 | 1.55 | 0.65
32 | 20.62 | 1.55 | 0.64
64 | 39.58 | 1.62 | 0.62
128 | 73.93 | 1.73 | 0.58
256 | 125.14 | 2.05 | 0.49
512 | 252.02 | 2.03 | 0.49
1k | 475.09 | 2.16 | 0.46
2k | 855.11 | 2.4 | 0.42
4k | 1433.35 | 2.86 | 0.35
RxM/verbs/MSG |   |   |  
bytes | MB/sec | usec/xfer | Mxfers/sec
1 | 0.96 | 1.04 | 0.96
2 | 1.93 | 1.04 | 0.97
4 | 3.86 | 1.04 | 0.97
8 | 7.72 | 1.04 | 0.97
16 | 15.3 | 1.05 | 0.96
32 | 30.54 | 1.05 | 0.95
64 | 59.36 | 1.08 | 0.93
128 | 114.92 | 1.11 | 0.9
256 | 167.69 | 1.53 | 0.66
512 | 310.28 | 1.65 | 0.61
1k | 572.5 | 1.79 | 0.56
2k | 964.13 | 2.12 | 0.47
4k | 1427.08 | 2.87 | 0.35

File: prov/verbs/src/fi_verbs.h [hunk 1]
```diff
@@ -789,4 +792,93 @@ fi_ibv_send_iov_flags(struct fi_ibv_ep *ep, struct ibv_send_wr *wr,
 	return fi_ibv_send_poll_cq_if_needed(ep, wr);
 }
 
+static inline const void *
+fi_ibv_dgram_moderate_user_buf(const void *user_buf, size_t offset)
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.h]: What does the term 'moderate' mean with respect to this set of calls?

**Comment by dmitrygladkov** [comment]
> "+466 −1,197" is always a good sign. :)

Yes, a lot of stupid code were removed :)

File: prov/verbs/src/fi_verbs.h (hunk 1 above)
Comment (dmitrygladkov) [reply prov/verbs/src/fi_verbs.h]: I'm not sure that's the term 'moderate' correct here. Do you have any suggestion?
I mean that the functions change user's buffer/iov to consider message prefix that's provided in the message (ibv doesn't require this for the send message -- that's why we rewind the user's buffer/iov --, it is only required for receive buffer)

**Comment by shefty** [comment]
@a-ilango - Did you want to review this prior to merging?  Or are you okay with just merging?  I didn't look too closely at the changes, but nothing jumped out at me.

**Comment by a-ilango** [comment]
Let me do a quick scan. Btw, how's verbs/MSG latency higher than RxM/verbs/MSG?

File: prov/verbs/src/verbs_domain.c [hunk 2]
```diff
@@ -60,12 +58,7 @@ static int fi_ibv_domain_bind(struct fid *fid, struct fid *bfid, uint64_t flags)
 			domain->eq_flags = flags;
 			break;
 		case FI_EP_DGRAM:
-			dgram_eq = container_of(bfid, struct fi_ibv_dgram_eq,
-						util_eq.eq_fid);
-			if (!dgram_eq)
-				return -FI_EINVAL;
-			return ofi_domain_bind_eq(&domain->util_domain,
-						  &dgram_eq->util_eq);
+			return -FI_EINVAL;
```
Comment (a-ilango) [inline prov/verbs/src/verbs_domain.c]: Is EQ for DGRAM endpoints being removed?

Comment (dmitrygladkov) [reply prov/verbs/src/verbs_domain.c]: Yes, EQ, CNTR has been removed, because they are not needed.
Once they are needed, we add them. Is it OK?

**Comment by shefty** [comment]
I'll guess that the test calls fi_send(), which verbs passes through to ibv_post_send, but rxm converts into an inject call, which ends up using inline data.

**Comment by dmitrygladkov** [comment]
> Btw, how's verbs/MSG latency higher than RxM/verbs/MSG?

I'm not considering fi_pingpong as a true benchmark, but this is due to SW optimizations on RxM level (pre-posted buffers)

File: prov/verbs/src/verbs_domain.c (hunk 2 above)
Comment (a-ilango) [reply prov/verbs/src/verbs_domain.c]: Sure. Please add this info in the commit message.

Comment (dmitrygladkov) [reply prov/verbs/src/verbs_domain.c]: > Please add this info in the commit message

Sure

**Comment by shefty** [comment]
@dmitrygladkov - please merge when ready

**Pull request merged**